libs/gltfio/src/MorphHelper.cpp

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MorphHelper.h"
+
+#include <filament/RenderableManager.h>
+#include <filament/VertexBuffer.h>
+
+#include "GltfEnums.h"
+#include "TangentsJob.h"
+
+using namespace filament;
+using namespace filament::math;
+using namespace utils;
+
+static const auto kFreeCallback = [](void* mem, size_t, void*) { free(mem); };
+static constexpr uint8_t kUnused = 0xff;
+
+namespace gltfio {
+
+uint32_t computeBindingSize(const cgltf_accessor* accessor);
+uint32_t computeBindingOffset(const cgltf_accessor* accessor);
+
+// Returns true if a is a subset of b.
+static bool isSubsetOf(ubyte4 a, ubyte4 b) {
+    if (a.x != kUnused && a.x != b.x && a.x != b.y && a.x != b.z && a.x != b.w) return false;
+    if (a.y != kUnused && a.y != b.x && a.y != b.y && a.y != b.z && a.y != b.w) return false;
+    if (a.z != kUnused && a.z != b.x && a.z != b.y && a.z != b.z && a.z != b.w) return false;
+    if (a.w != kUnused && a.w != b.x && a.w != b.y && a.w != b.z && a.w != b.w) return false;
+    return true;
+}
+
+MorphHelper::MorphHelper(FFilamentAsset* asset, FFilamentInstance* inst) : mAsset(asset),
+        mInstance(inst) {
+    NodeMap& sourceNodes = asset->isInstanced() ? asset->mInstances[0]->nodeMap : asset->mNodeMap;
+    for (auto pair : sourceNodes) {
+        mMorphTable[pair.second].sourceNode = pair.first;
+    }
+}
+
+MorphHelper::~MorphHelper() {
+    auto engine = mAsset->mEngine;
+    for (auto& entry : mMorphTable) {
+        for (auto& perm : entry.second.permutations) {
+            if (perm.owner) {
+                for (auto& prim : perm.primitives) {
+                    engine->destroy(prim.vertices);
+                }
+            }
+        }
+    }
+}
+
+void MorphHelper::applyWeights(Entity entity, float const* weights, size_t count) {
+    auto renderableManager = &mAsset->mEngine->getRenderableManager();
+    auto renderable = renderableManager->getInstance(entity);
+
+    // If there are 4 or fewer targets, we can simply re-use the original VertexBuffer.
+    if (count <= 4) {
+        if (mEnabled) {
+            float4 vec(0, 0, 0, 0);
```

Comment:
`float4 vec{}` -- slightly less verbose, not a biggie.

Reply:
Don.e